File: src/main/java/com/exampledigisphere/vitapetcare/admin/file/domain/File.java
package com.exampledigisphere.vitapetcare.admin.file.domain;

import com.exampledigisphere.vitapetcare.admin.user.domain.User;
import com.exampledigisphere.vitapetcare.config.root.BaseEntity;
import com.fasterxml.jackson.annotation.JsonView;
import jakarta.persistence.*;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;
import org.hibernate.Hibernate;

import java.util.Objects;

@Entity
@Table(name = "fil_files")
public class File extends BaseEntity {

  @Transient
  final String TABLE_NAME = "fil_files";

  @NotBlank
  @JsonView(Json.List.class)
  private String name;

  @NotBlank
  @JsonView(Json.List.class)
  private String path;

  @NotNull
  @ManyToOne(fetch = FetchType.LAZY)
  @JoinColumn(name = "user_id")
  private User user;

  @NotNull
  @Enumerated(EnumType.STRING)
  @JsonView(Json.Detail.class)
  private FileType type;

  public File() {
  }

  public String getName() {
    return name;
  }

  public void setName(String name) {
    this.name = name;
  }

  public String getPath() {
    return path;
  }

  public void setPath(String path) {
    this.path = path;
  }

  public User getUser() {
    return user;
  }

  public void setUser(User user) {
    this.user = user;
  }

  public FileType getType() {
    return type;
  }

  public void setType(FileType type) {
    this.type = type;
  }

  public File loadUser() {
    Hibernate.initialize(user);
    return this;
  }

  public interface Authority {
    String FILE_CREATE = "FILE_CREATE";
    String FILE_EDIT = "FILE_EDIT";
    String FILE_VIEW = "FILE_VIEW";
    String FILE_VIEW_LIST = "FILE_VIEW_LIST";
    String FILE_DELETE = "FILE_DELETE";
  }

  @Override
  public boolean equals(Object object) {
    if (object == null || getClass() != object.getClass()) return false;
    if (!super.equals(object)) return false;
    File file = (File) object;
    return Objects.equals(name, file.name)
      && Objects.equals(path, file.path)
      && Objects.equals(user, file.user)
      && type == file.type;
  }

  @Override
  public int hashCode() {
    return Objects.hash(super.hashCode(), TABLE_NAME, name, path, user, type);
  }
}

File: src/main/kotlin/com/exampledigisphere/vitapetcare/admin/file/domain/File.kt
package com.exampledigisphere.vitapetcare.admin.file.domain

import com.exampledigisphere.vitapetcare.admin.user.domain.User
import com.exampledigisphere.vitapetcare.config.root.BaseEntity
import com.exampledigisphere.vitapetcare.config.security.Permissions
import com.fasterxml.jackson.annotation.JsonIgnoreProperties
import com.fasterxml.jackson.annotation.JsonView
import jakarta.persistence.*
import jakarta.validation.constraints.NotBlank
import org.hibernate.Hibernate
import org.jetbrains.annotations.NotNull

@Entity
@Table(name = "fil_files")
class File : BaseEntity() {

  @field:Transient
  final val TABLE_NAME = "fil_files";

  @field:NotBlank
  @field:JsonView(Json.List::class)
  var name: String? = null;

  @field:NotBlank
  @field:JsonView(Json.List::class)
  var path: String? = null;

  @field:NotNull
  @ManyToOne(fetch = FetchType.LAZY)
  @field:JoinColumn(name = "user_id")
  @field:JsonView(Json.WithUser::class)
  @field:JsonIgnoreProperties("user")
  var user: User? = null;

  @field:NotNull
  @field:Enumerated(EnumType.STRING)
  @field:JsonView(Json.Detail::class)
  var type: FileType? = null;

  fun loadOwner() = Hibernate.initialize(user);

  interface Json {
    interface List;
    interface Detail : List;
    interface WithUser : User.Json.List
    interface All : Detail, WithUser;
  }

  object FilePermissions : Permissions {
    override var resource = "FILE"
  }
}
